src/dict.c

```diff
@@ -870,136 +869,163 @@ void dictTwoPhaseUnlinkFree(dict *d, dictEntry *he, dictEntry **plink, int table
     dictResumeRehashing(d);
 }
 
+#define DICT_SET_KEY(e, k)                                                                                             \
```

Comment:
Because the field can be of different types so we will need to write a new function for every type without macros (or templates). This is essentially "template metaprogramming" in (macros).

Reply:
For DICT_SET_KEY, that's not true, I believe. The args are just values, so it can be a function.

For the others DICT_SET_VAL, instead of taking a field name as a parameter, we can let it be a function that returns a pointer to the value union within the entry.

It can be called like

```c
    entryValue(de)->u64 = val;
```

Then it can be a static inline function.

We need to define the value-union type outside the entry structs too, so we can return a pointer to this union type. (Before the embedded entry, only the normal entey had value, so there was no need for that.)

Reply:
Yeah, I would prefer DICT_SET_KEY be an inline function as well.

Reply:
Me too. Also, debug-ability with function is much better than macro.

Reply:
Draft (untested):

```C
union value {
    void *val;
    uint64_t u64;
    int64_t s64;
    double d;
};

typedef struct {
    union value v;
    struct dictEntry *next;  /* Next entry in the same hash bucket. */
    uint8_t key_header_size; /* offset into key_buf where the key is located at. */
    unsigned char key_buf[]; /* buffer with embedded key. */
} dictEntryEmbedded;

static inline union value *dictEntryValuePtr(dictEntry *de) {
    if (entryIsNormal(de)) {
        dictEntryNormal *_de = decodeEntryNormal(de);
        return &_de->v;
    } else if (entryIsEmbedded(de)) {
        dictEntryEmbedded *_de = decodeEntryEmbedded(de);
        return &_de->v;
    } else {
        assert(0);
    }
}


```

Reply:
This union defeats the purpose of reducing the scope of type casting because the caller will now need to track additional context in order to select the correct field to access. The distance of the context and the value is increased hence making this even more error prone. Even worse, it actually gives the core review false safety since there is no casting on appearance. I will continue with the macros.

Reply:
It's not a new union. It was an unnamed union, identical in two different structs. 🤷

Reply:
discussed offline with @madolson. converted `DICT_SET_KEY` to `dictSetEntryKey`